types/nexelec/origin/uplink.js

```diff
@@ -0,0 +1,269 @@
+    return;
+  }
+  emitIfNotEmpty("system", { productType }, timestamp);
+
+  const climate = {};
+  climate.minTemperature = decodeTemperature((payload[2] << 2) | (payload[3] >> 6));
+  climate.maxTemperature = decodeTemperature(((payload[3] & 0x3f) << 4) | (payload[4] >> 4));
+  climate.averageTemperature = decodeTemperature(((payload[4] & 0x0f) << 6) | (payload[5] >> 2));
+  climate.minHumidity = decodeHumidity(((payload[5] & 0x03) << 6) | (payload[6] >> 2));
+  climate.maxHumidity = decodeHumidity(((payload[6] & 0x03) << 6) | (payload[7] >> 2));
+  climate.averageHumidity = decodeHumidity(((payload[7] & 0x03) << 6) | (payload[8] >> 2));
+
+  emitIfNotEmpty("default", climate, timestamp);
+}
+
+// Handles Periodic Data messages (0x04).
+function handlePeriodicData(payload, productType, timestamp) {
+  if (payload.length < 5) {
+    emit("log", { error: "Invalid payload length for Periodic Data" });
+    return;
+  }
+  emitIfNotEmpty("system", { productType }, timestamp);
+
+  const climate = {};
+  climate.temperature = decodeTemperature((payload[2] << 2) | (payload[3] >> 6));
+  climate.humidity = decodeHumidity(((payload[3] & 0x3f) << 2) | (payload[4] >> 6));
+
+  emitIfNotEmpty("default", climate, timestamp);
+}
+
+// Handles Historical Data messages (0x05).
+function handleHistoricalData(payload, productType, timestamp) {
+  emitIfNotEmpty("system", { productType }, timestamp);
+
+  const datalog = {};
+  datalog.totalMeasurements = payload[2] >> 2;
+  datalog.samplingPeriod = ((payload[2] & 0x03) << 6) | (payload[3] >> 2);
+  datalog.repetitionsInMessage = ((payload[3] & 0x03) << 4) | (payload[4] >> 4);
+  emitIfNotEmpty("datalog", datalog, timestamp);
+
+  const baseTimeMs = new Date(timestamp).getTime();
+  const periodMs = datalog.samplingPeriod * 60 * 1000;
+
+  let bitString = "";
+  for (let i = 0; i < payload.length; i++) {
+    bitString += payload[i].toString(2).padStart(8, '0');
+  }
+
+  for (let i = 0; i < datalog.totalMeasurements; i++) {
+    const startBit = 36 + (i * 10);
+    const endBit = startBit + 10;
+    if (endBit > bitString.length) break;
+
+    const tempBits = bitString.substring(startBit, endBit);
+    const tempRaw = parseInt(tempBits, 2);
+    const temperature = decodeTemperature(tempRaw);
+
+    if (temperature !== null) {
+      const measurementTimeMs = baseTimeMs - (i * periodMs);
+      const measurementTimestamp = new Date(measurementTimeMs).toISOString();
+      emitIfNotEmpty("default", { temperature }, measurementTimestamp);
+    }
+  }
+}
+
+
+/**
+ * Decodes the payload from a nexelec origin device.
+ * @param {object} event - The event object containing the payload and other data.
+ */
+function consume(event) {
+  const payload = Hex.hexToBytes(event.data.payloadHex);
+  const { port } = event.data;
+
+  // Create a reliable timestamp, falling back to the current time if invalid.
+  let reliableTimestamp = event.time;
```

Comment:
this logic is not required as event.timestamp will be used anyways if not set

note that there is no event.time property, where did you find this? it is timestamp instead